client/tests/integration/smartcontracts/validator_with_custom_token/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+        visit_burn_account_public_key(Burn<Account, PublicKey>),
+        visit_mint_account_signature_check_condition(Mint<Account, SignatureCheckCondition>),
+        visit_set_account_key_value(SetKeyValue<Account>),
+        visit_remove_account_key_value(RemoveKeyValue<Account>),
+
+        // Asset validation
+        visit_register_asset(Register<Asset>),
+        visit_unregister_asset(Unregister<Asset>),
+        visit_mint_asset(Mint<Asset, NumericValue>),
+        visit_burn_asset(Burn<Asset, NumericValue>),
+        visit_transfer_asset(Transfer<Asset, NumericValue, Account>),
+        visit_set_asset_key_value(SetKeyValue<Asset>),
+        visit_remove_asset_key_value(RemoveKeyValue<Asset>),
+
+        // AssetDefinition validation
+        visit_unregister_asset_definition(Unregister<AssetDefinition>),
+        visit_transfer_asset_definition(Transfer<Account, AssetDefinition, Account>),
+        visit_set_asset_definition_key_value(SetKeyValue<AssetDefinition>),
+        visit_remove_asset_definition_key_value(RemoveKeyValue<AssetDefinition>),
+
+        // Permission validation
+        visit_register_permission_token(Register<PermissionTokenDefinition>),
+        visit_grant_account_permission(Grant<Account, PermissionToken>),
+        visit_revoke_account_permission(Revoke<Account, PermissionToken>),
+
+        // Role validation
+        visit_unregister_role(Unregister<Role>),
+        visit_grant_account_role(Grant<Account, RoleId>),
+        visit_revoke_account_role(Revoke<Account, RoleId>),
+
+        // Trigger validation
+        visit_unregister_trigger(Unregister<Trigger<FilterBox, Executable>>),
+        visit_mint_trigger_repetitions(Mint<Trigger<FilterBox, Executable>, u32>),
+        visit_execute_trigger(ExecuteTrigger),
+
+        // Parameter validation
+        visit_set_parameter(SetParameter),
+        visit_new_parameter(NewParameter),
+
+        // Upgrade validation
+        visit_upgrade_validator(Upgrade<iroha_validator::data_model::validator::Validator>),
+    }
+}
+
+impl Validate for CustomValidator {
+    fn verdict(&self) -> &Result {
+        self.0.verdict()
+    }
+    fn deny(&mut self, reason: ValidationFail) {
+        self.0.deny(reason);
+    }
+}
+
+impl ExpressionEvaluator for CustomValidator {
+    fn evaluate<E: Evaluate>(
+        &self,
+        expression: &E,
+    ) -> core::result::Result<E::Value, EvaluationError> {
+        self.0.evaluate(expression)
+    }
+}
+
+/// Entrypoint to return permission token definitions defined in this validator.
+#[entrypoint]
+pub fn permission_tokens() -> Vec<PermissionTokenDefinition> {
+    let mut tokens = iroha_validator::default::permission_tokens();
+
+    // TODO: Not very convenient usage.
+    // We need to come up with a better way.
+    if let Some(pos) = tokens.iter().position(|definition| {
+        definition == &iroha_validator::default::domain::tokens::CanUnregisterDomain::definition()
+    }) {
+        tokens.remove(pos);
+    }
```

Comment:
Why no `retain`?

Reply:
Because `retain()` will iterate over all collection, while `position()` will stop on the first found result

Reply:
Do you expect size of permission tokens be that big that it will make any difference?

Reply:
ok, i just thought that it will be more compact
feel free to resolve

Reply:
> ok, i just thought that it will be more compact
feel free to resolve

Appreciate your opinion, but, although as Marin said I win only some comparisions, I'd like to keep it this way. Just a feeling 😄

Reply:
`swap_remove` will do the job if we don't care about order of permission tokens